### ISSUE TITLE
Overwriting a record entry causes an esoteric type error

### ISSUE THREAD

**Issue by euclidianAce** [teal-language/tl#167]
~~Basically function signatures are not compared to the signature of functiontypes, just their "names" are. This causes problems attempting to implement a function of a type that a library provides (such as ltn12)~~.

Some of the errors produced in the following examples seem esoteric and misleading.

The following code produces a seemingly incorrect error message
```
-- foo.tl

local foo = record
	bar = functiontype<T>(T)
	example: bar<string>
end

function foo.example(data: string)
	print(data)
end

```
```
$ tl check foo.tl
========================================
1 error:
foo.tl:5:11: unknown type bar<string>

```
~~But the real issue is that the signature of the implementation and declaration(?) don't align as taking the implementation away passes the check.~~

Attempting to annotate the declaration also produces an error
```
-- foo.tl

local foo = record
	bar = functiontype<T>(T)
	example: bar<string>
end

foo.example: bar<string> = function(data: string)
	print(data)
end
```
```
$ tl check foo.tl
========================================
2 syntax errors:
foo.tl:8:26: syntax error
foo.tl:8:28: syntax error
```
And as a last resort, casting the function to force the type inference results in an even worse error
```
-- foo.tl

local foo = record
	bar = functiontype<T>(T)
	example: bar<string>
end

foo.example = function(data: string)
	print(data)
end as bar<string>
```
```
$ tl check foo.tl
========================================
2 errors:
foo.tl:5:11: unknown type bar<string>
foo.tl:10:8: in assignment: bar<string> is not a bar<string>
```
I ran into this issue writing something where I wanted the types of some functions to be Filters and Sinks from ltn12, and there doesn't seem to be a way to annotate a function like this.

**Comment by pdesaulniers** [comment]
Looking at this error in your first example:

```lua
foo.tl:5:11: unknown type bar<string>
```

Maybe you just need to specify `foo.bar<string>` at line 5? The following appears to work:

```lua
-- foo.tl

local foo = record
	bar = functiontype<T>(T)
	example: foo.bar<string>
end

return foo
```

```lua
-- test.tl

local foo = require("foo")

function foo.example(data: string)
	print(data)
end
```

**Comment by pdesaulniers** [comment]
> But the real issue is that the signature of the implementation and declaration(?) don't align as taking the implementation away passes the check.

In your example above:

```lua
local foo = record
	bar = functiontype<T>(T)
	example: foo.bar<string>
end

function foo.example(data: string)
	print(data)
end
```

I believe you're seeing the effects of the 'record extension feature' (for the lack of a better name...)

This feature allows adding methods to records, as long as they are defined in the same lexical scope as the record declaration:

```lua
local Point = record
   x: number
   y: number
end

function Point.new(x: number, y: number): Point
   -- return something
end
```

In your example, I believe you're actually overriding the type of `foo.example`, because the implementation is in the same scope as the record declaration.

**Comment by euclidianAce** [comment]
Okay, that makes sense. Some of those type errors are still a little strange, and I guess what I'm really asking for is some syntactic sugar for functiontype of functions where I could just annotate a function with a functiontype
```
local function: ltn12.Sink<string> foo(data)
    -- ...
end
```
Or something to that effect to not have to rewrite the function signature when it's already aliased

**Comment by euclidianAce** [comment]
I feel like I should also mention that this was caused by me trying to implement a method that I had accidentally put in a record and also left an empty function to fill in later. This caused a type error in the ltn12.d.tl file rather than my module. A reduced example of that:
```
local ltn12 = require("ltn12")
local foo = record
   send: ltn12.Pump<string> -- forgot how extending records worked
end

-- [[ ... ]]

function foo.send(data: string)
   -- ...
end
```
This resulted in a similar error as above, but in the ltn12.d.tl file, saying that `Sink<string>` wasn't a type at the `Pump` definition.
`[...]/ltn12.d.tl:14:38: unknown type Sink<string>`

**Comment by hishamhm** [comment]
@euclidianAce Thank you for all the examples and also for providing context! These error messages definitely need to be fixed... and quite frankly, the very first thing you tried (very first `foo.tl` in the original message) should have worked, because those nested types _sort of_ work for records as of current `master` (98b818):

```
-- no errors!
local foo = record
   bar = record<T>
      x: T
   end
   example: bar<string>
end
```

but then

```
-- fails!? o_O
local foo = record
   bar = record<T>
      x: T
   end
   example: bar<string>
end

local f: foo = {}

foo.example = { x = "hello" }
```
produces
```
========================================
2 errors:
foo.tl:6:13: unknown type bar<string>
foo.tl:11:15: in assignment: got {x: string "hello"}, expected bar<string>
```

**Comment by hishamhm** [comment]
and then if you apply @pdesaulniers 's "trick" using the full path of the name, it works again:

```
-- no errors
local foo = record
   bar = record<T>
      x: T
   end
   example: foo.bar<string> -- <=== full path foo.bar
end

local f: foo = {}

foo.example = { x = "hello" }
```

and we can see that it is type checking for real: the expected errors occur if you misuse `f.example`:

```
local foo = record
   bar = record<T>
      x: T
   end
   example: foo.bar<string> -- <=== full path foo.bar
end

local f: foo = {}

f.example = { x = 12, y = 9 }
```

produces the correct errors:
```
========================================
2 errors:
foo.tl:11:19: in assignment: record field doesn't match: x: got number, expected string
foo.tl:11:27: in assignment: unknown field y
```

**Comment by hishamhm** [comment]
Wow, I spent quite some time trying to track this down on Saturday and after a night of sleep and a clearer mind, I found a very clean solution which solved not only these but also other related problems (see the regression tests in commit 149859d!)

Really happy with the result — this ended up cleaning up a few spots in the code base which were very clunky and really annoyed me. For those types of algorithms, it's always a very good sign when the fixed is actually simpler!

@euclidianAce Now your first example works:

```
local foo = record
   bar = functiontype<T>(T)
   example: bar<string>
end

function foo.example(data: string)
   print(data)
end
```

And the one with the cast actually works too if you give it the full type path:

```
local foo = record
	bar = functiontype<T>(T)
	example: bar<string>
end

foo.example = function(data: string)
	print(data)
end as foo.bar<string>
```

(and it does _not_ give an esoteric error if you just do `as bar<string>`!)

Thank you so much for the reduced test cases!
